controllers/constraintstatus_controller.go

```diff
@@ -0,0 +1,274 @@
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var ControllerName = "contraintstatus_reconciler"
+
+type discoveryInfo struct {
+	apiResourceList        []*metav1.APIResourceList
+	discoveryLastRefreshed time.Time
+}
+
+type ConstraintStatusReconciler struct {
+	client.Client
+	Log                     logr.Logger
+	DynamicClient           dynamic.Interface
+	ClientSet               *kubernetes.Clientset
+	Scheme                  *runtime.Scheme
+	Namespace               string
+	MaxConcurrentReconciles uint
+	discoveryInfo
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ConstraintStatusReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		// The work queue prevents the same item being reconciled concurrently:
+		// https://github.com/kubernetes-sigs/controller-runtime/issues/1416#issuecomment-899833144
+		WithOptions(controller.Options{MaxConcurrentReconciles: int(r.MaxConcurrentReconciles)}).
+		// watch Config resrouce as secondary
+		Owns(&v1alpha1.Config{}).
+		Named(ControllerName).
+		For(&v1beta1.ConstraintPodStatus{},
+			builder.WithPredicates(predicate.Funcs{
+				// Execute this reconcile func when it is audit-constraintStatuspod
+				// because a constraint creates 4 constraintStatuspods
+				CreateFunc: func(e event.CreateEvent) bool {
+					obj := e.Object.(*v1beta1.ConstraintPodStatus)
+
+					return slices.Contains(obj.Status.Operations, "audit")
+				},
+				UpdateFunc: func(e event.UpdateEvent) bool {
+					oldObj := e.ObjectOld.(*v1beta1.ConstraintPodStatus)
+					newObj := e.ObjectNew.(*v1beta1.ConstraintPodStatus)
+
+					return slices.Contains(newObj.Status.Operations, "audit") &&
+						oldObj.Status.ObservedGeneration != newObj.Status.ObservedGeneration
+				},
+				DeleteFunc: func(e event.DeleteEvent) bool {
+					obj := e.Object.(*v1beta1.ConstraintPodStatus)
+
+					return slices.Contains(obj.Status.Operations, "audit")
+				},
+			},
+			)).
+		Complete(r)
+}
+
+// User set gatekeeper.spec.audit.auditFromCache to Automatic, this reconcile function
+// collect all constraints and add found kinds which is used in contraint to config.spec.sync.syncOnly
+func (r *ConstraintStatusReconciler) Reconcile(ctx context.Context,
+	request reconcile.Request,
+) (reconcile.Result, error) {
+	log := r.Log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	log.Info("Reconciling ConstraintPodStatus and Config")
+
+	gatekeeper := &operatorv1alpha1.Gatekeeper{}
+	err := r.Get(ctx, types.NamespacedName{
+		Namespace: "",
+		Name:      "gatekeeper",
+	}, gatekeeper)
+
+	auditFromCache := *gatekeeper.Spec.Audit.AuditFromCache
+
+	if err != nil || !strings.EqualFold(string(auditFromCache), "Automatic") {
+		return reconcile.Result{Requeue: false}, fmt.Errorf("auditFromCache is not set so skip this reconcile")
```

Comment:
This shouldn't return an error because even if requeue is false, an error causes a requeue of the request.